providers/implementations/ciphers/cipher_enull_hmac.c

```diff
@@ -0,0 +1,343 @@
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_einit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+        ret = hw->initiv(ctx, iv, ivlen);
+    }
+    if (ret && !enull_hmac_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int enull_hmac_dinit(void *vctx,
+                            const unsigned char *key, size_t keylen,
+                            const unsigned char *iv, size_t ivlen,
+                            const OSSL_PARAM params[])
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_dinit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+        hw->initiv(ctx, iv, ivlen);
+    }
+    if (ret && !enull_hmac_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int enull_hmac_cipher(void *vctx, unsigned char *out,
+                             size_t *outl, size_t outsize,
+                             const unsigned char *in, size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (inl == 0) {
+        *outl = 0;
+        return 1;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!hw->base.cipher(ctx, out, in, inl))
+        return 0;
+    *outl = inl;
+    return 1;
+}
+
+static int enull_hmac_final(void *vctx, unsigned char *out, size_t *outl,
+                            size_t outsize)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!hw->base.cipher(ctx, out, NULL, 0))
+        return 0;
+
+    *outl = 0;
+    return 1;
+}
+
+#define IMPLEMENT_cipher(lcmd, UCMD, flags, kbits, blkbits, ivbits)            \
+static OSSL_FUNC_cipher_get_params_fn enull_hmac_##lcmd##_get_params;          \
+static int enull_hmac_##lcmd##_get_params(OSSL_PARAM params[])                 \
+{                                                                              \
+    return ossl_cipher_generic_get_params(params, 0, flags,                    \
+                                          kbits, blkbits, ivbits);             \
+}                                                                              \
+                                                                               \
+static OSSL_FUNC_cipher_newctx_fn enull_hmac_##lcmd##_newctx;                  \
```

Comment:
Why this forward declaration, given the following line actually defining enull_hmac_##lcmd##_newctx?

Reply:
It is not forward declaration, `OSSL_FUNC_cipher_newctx_fn ` is macro for typedef function signature.

Reply:
yes, It is just function signature, can be removed. But I think its a way to enforce function signature.